pallets/messages/src/lib.rs

```diff
@@ -315,28 +319,26 @@ impl<T: Config> Pallet<T> {
 		msa_id: Option<MessageSourceId>,
 		payload: BoundedVec<u8, T::MaxMessagePayloadSizeBytes>,
 		schema_id: SchemaId,
-	) -> Result<Message<T::MaxMessagePayloadSizeBytes>, DispatchError> {
-		<BlockMessages<T>>::try_mutate(
-			|existing_messages| -> Result<Message<T::MaxMessagePayloadSizeBytes>, DispatchError> {
-				let current_size: u16 = existing_messages
-					.len()
-					.try_into()
-					.map_err(|_| Error::<T>::TypeConversionOverflow)?;
-
-				let msg = Message {
-					payload, // size is checked on top of extrinsic
-					provider_msa_id,
-					msa_id,
-					index: current_size,
-				};
-
-				existing_messages
-					.try_push((msg.clone(), schema_id))
-					.map_err(|_| Error::<T>::TooManyMessagesInBlock)?;
-
-				Ok(msg)
-			},
-		)
+		current_block: T::BlockNumber,
+	) -> DispatchResult {
+		<Messages<T>>::try_mutate(current_block, schema_id, |existing_messages| -> DispatchResult {
+			let extrinsic_index =
+				<frame_system::Pallet<T>>::extrinsic_index().ok_or(Error::<T>::BadContext)?;
+			let msg = Message {
+				payload, // size is checked on top of extrinsic
+				provider_msa_id,
+				msa_id,
+				// converting u32 to u16 and defaulting to u16:MAX if overflowed
+				index: extrinsic_index.try_into().unwrap_or(u16::MAX),
```

Comment:
the probability of having more than 65,536 extrinsics in a block is slim to none but just in case something happens we will use max value for our sorting